File: 2025-08-August-GFG-POTD/August 02 - Longest Subarray with Majority Greater than K/Solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int longestSubarray(vector<int>& arr, int k) {
        int n = arr.size();
        vector<int> transformed(n);

        // Step 1: Transform the array
        for (int i = 0; i < n; ++i) {
            transformed[i] = (arr[i] > k) ? 1 : -1;
        }

        // Step 2: Use prefix sum and hashmap
        unordered_map<int, int> firstIndex;
        int maxLen = 0;
        int prefixSum = 0;

        for (int i = 0; i < n; ++i) {
            prefixSum += transformed[i];

            if (prefixSum > 0) {
                maxLen = i + 1;
            } else {
                if (firstIndex.find(prefixSum) == firstIndex.end()) {
                    firstIndex[prefixSum] = i;
                }

                if (firstIndex.find(prefixSum - 1) != firstIndex.end()) {
                    int prevIndex = firstIndex[prefixSum - 1];
                    maxLen = max(maxLen, i - prevIndex);
                }
            }
        }

        return maxLen;
    }
};

File: 2025-08-August-GFG-POTD/August 02 - Longest Subarray with Majority Greater than K/Solution.java
class Solution {
    public int longestSubarray(int[] arr, int k) {
        int n = arr.length;
        int[] transformed = new int[n];

        // Step 1: Transform the array
        for (int i = 0; i < n; i++) {
            transformed[i] = (arr[i] > k) ? 1 : -1;
        }

        // Step 2: Use prefix sum and hashmap
        HashMap<Integer, Integer> firstIndex = new HashMap<>();
        int maxLen = 0;
        int prefixSum = 0;

        for (int i = 0; i < n; i++) {
            prefixSum += transformed[i];

            // Case 1: If prefixSum is positive till this point
            if (prefixSum > 0) {
                maxLen = i + 1;
            }
            // Case 2: If prefixSum is seen before, try to find a longer subarray
            else {
                if (!firstIndex.containsKey(prefixSum)) {
                    firstIndex.put(prefixSum, i);
                }

                // Check if (prefixSum - 1) was seen before
                if (firstIndex.containsKey(prefixSum - 1)) {
                    int prevIndex = firstIndex.get(prefixSum - 1);
                    maxLen = Math.max(maxLen, i - prevIndex);
                }
            }
        }

        return maxLen;
    }
}
